### PR TITLE
graph-struct-0: initial project ideas

### PR DESCRIPTION
**Pull request by thequickf**

### CODE REVIEW

File: src/edge.hpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef EDGE_HPP
+#define EDGE_HPP
+
+namespace graph {
+
+template<typename Node, typename... Types>
+class Edge : public Types... {
+public:
+    Edge(const Node& from, const Node& to, Types... args) : from(from), to(to), Types(args)... {}
```
Comment (isnullxbh) [inline src/edge.hpp]: Pass by forwarding references and use `std::forward`:

`Types&&... args` => `Types(std::forward<Ts>(args))`

`from`, `to` - pass by value and use `std::move`

File: src/graph.hpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#ifndef GRAPH_HPP
+#define GRAPH_HPP
+
+#include <edge.hpp>
+
+namespace graph {
+
+template<typename Node, typename... Types>
```
Comment (isnullxbh) [inline src/graph.hpp:8]: Make `Edge` a template parameter.

File: src/impl/directed_graph.hpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <edge.hpp>
+#include <graph.hpp>
+
+#include <map>
+#include <set>
+
+namespace graph {
+
+template<typename Node, typename... Types>
```
Comment (isnullxbh) [inline src/impl/directed_graph.hpp:12]: Move the definition to the edge.hpp

File: src/edge.hpp (hunk 1 above)
Comment (thequickf) [reply src/edge.hpp]: Why would we force users to use ```std::move```?
```Types&&... args => Types(std::forward<Ts>(args))```

File: src/impl/directed_graph.hpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+bool operator<(const Edge<Node, Types...>& lhe, const Edge<Node, Types...>& rhe) {
+    if (lhe.from != rhe.from)
+        return lhe.from < rhe.from;
+    return lhe.to < rhe.to;
+}
+
+template<typename Node, typename... Types>
+class DirectedGraph : Graph<Node, Types...> {
```
Comment (isnullxbh) [inline src/impl/directed_graph.hpp]: public Graph<Node, Types...>?

File: src/edge.hpp (hunk 1 above)
Comment (thequickf) [reply src/edge.hpp]: Fixed

File: src/impl/directed_graph.hpp (hunk 3 above)
Comment (thequickf) [reply src/impl/directed_graph.hpp:12]: moved ```operator<``` for now

File: src/impl/directed_graph.hpp (hunk 4 above)
Comment (thequickf) [reply src/impl/directed_graph.hpp]: Thanks. Fixed

File: src/graph.hpp (hunk 2 above)
Comment (thequickf) [reply src/graph.hpp:8]: Can you make PR please?

**Pull request merged**